staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +906,215 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
 	m.succeedCount = len(txns)
 	return nil
 }
+
+type eventRecord struct {
+	Action string
+	Key    string
+}
+
+type mockReplaceTester struct {
+	atomic     bool
+	source     *fcache.FakeControllerSource
+	store      Store
+	controller Controller
+
+	historyLock sync.Mutex
+	history     []eventRecord
+}
+
+func setupMockReplaceTester(t *testing.T, atomic bool) *mockReplaceTester {
+	clientgofeaturetesting.SetFeatureDuringTest(t, clientgofeaturegate.AtomicFIFO, atomic)
+	source := fcache.NewFakeControllerSource()
+	store := NewStore(DeletionHandlingMetaNamespaceKeyFunc)
+
+	m := &mockReplaceTester{
+		atomic: atomic,
+		source: source,
+		store:  store,
+	}
+
+	record := func(action string, obj interface{}) {
+		m.historyLock.Lock()
+		defer m.historyLock.Unlock()
+		key, _ := DeletionHandlingMetaNamespaceKeyFunc(obj)
+		m.history = append(m.history, eventRecord{Action: action, Key: key})
+	}
+
+	handler := ResourceEventHandlerDetailedFuncs{
+		AddFunc: func(obj interface{}, isInInitialList bool) {
+			record("add", obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			record("update", newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			record("delete", obj)
+		},
+	}
+
+	c := newInformer(store, InformerOptions{
+		ListerWatcher: source,
+		ObjectType:    &v1.Pod{},
+		ResyncPeriod:  0,
+		Handler:       handler,
+	})
+	m.controller = c
+	return m
+}
+
+func (m *mockReplaceTester) teardown() {
+	m.source.Shutdown()
+}
+
+func (m *mockReplaceTester) clearHistory() {
+	m.historyLock.Lock()
+	defer m.historyLock.Unlock()
+	m.history = []eventRecord{}
+}
+
+func (m *mockReplaceTester) getHistory() []eventRecord {
+	m.historyLock.Lock()
+	historyCopy := make([]eventRecord, len(m.history))
+	copy(historyCopy, m.history)
+	m.historyLock.Unlock()
+
+	sortEvents := func(events []eventRecord) {
+		sort.Slice(events, func(i, j int) bool {
+			if events[i].Key != events[j].Key {
+				return events[i].Key < events[j].Key
+			}
+			return events[i].Action < events[j].Action
+		})
+	}
+	sortEvents(historyCopy)
+
+	return historyCopy
+}
+
+func (m *mockReplaceTester) getStoredItems() []interface{} {
+	items := m.store.List()
+
+	sortItems := func(items []interface{}) {
+		sort.Slice(items, func(i, j int) bool {
+			keyI, _ := DeletionHandlingMetaNamespaceKeyFunc(items[i])
+			keyJ, _ := DeletionHandlingMetaNamespaceKeyFunc(items[j])
+			return keyI < keyJ
+		})
+	}
+
+	sortItems(items)
+	return items
+}
+
+func (m *mockReplaceTester) runScenario(t *testing.T, ctx context.Context) {
```

Comment:
Running scenario is not property of tester. Test should own the test scenario. Tester is a helper to execute the test.